Add apache renegotiation test server to CI

goatgoose opened the pull request:
### Resolved issues:

None

### Description of changes: 

This PR adds an apache web server to the CI that will be used to test the s2n-tls renegotiation feature. The apache server hosts two endpoints: `/change_cipher_suite`, which forces a renegotiation to change the cipher suite, and `/mutual_auth`, which forces a renegotiation to request client authentication.

### Call-outs:

The apache server is installed and run locally, separate from the system's apache instillation, to ensure that customer's apache configurations are not overridden upon running our integration tests.

The integration tests for this PR will fail until the PR is approved and the docker image in CI is rebuilt with the changes to /codebuild.

### Testing:

It has been confirmed locally that both endpoints cause a renegotiation.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

goatgoose commented:
Codebuild run: https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/S2nIntegrationV2Batch2/batch/S2nIntegrationV2Batch2%3A248d786f-6315-46f3-b454-e8d6d3433212?region=us-west-2

lrstewart commented on codebuild/bin/install_ubuntu_dependencies.sh:
```diff
@@ -41,7 +41,7 @@ base_packages() {
   apt-get update -o Acquire::CompressionTypes::Order::=gz
 
-  DEPENDENCIES="unzip make indent iproute2 kwstyle libssl-dev net-tools  tcpdump valgrind lcov m4 nettle-dev nettle-bin pkg-config psmisc gcc g++ zlibc zlib1g-dev python3-pip python3-testresources llvm curl shellcheck git tox cmake libtool ninja-build golang-go quilt jq"
+  DEPENDENCIES="unzip make indent iproute2 kwstyle libssl-dev net-tools  tcpdump valgrind lcov m4 nettle-dev nettle-bin pkg-config psmisc gcc g++ zlibc zlib1g-dev python3-pip python3-testresources llvm curl shellcheck git tox cmake libtool ninja-build golang-go quilt jq apache2"
   if [[ -n "${GCC_VERSION:-}" ]] && [[ "${GCC_VERSION:-}" != "NONE" ]]; then
```
Super nit. It's just been bothering me for a while, and you are touching the line :)
```suggestion
  DEPENDENCIES="unzip make indent iproute2 kwstyle libssl-dev net-tools tcpdump valgrind lcov m4 nettle-dev nettle-bin pkg-config psmisc gcc g++ zlibc zlib1g-dev python3-pip python3-testresources llvm curl shellcheck git tox cmake libtool ninja-build golang-go quilt jq apache2"
  if [[ -n "${GCC_VERSION:-}" ]] && [[ "${GCC_VERSION:-}" != "NONE" ]]; then
```

lrstewart commented on codebuild/bin/apache2/ports.conf:
```diff
@@ -0,0 +1,7 @@
+# If you just change the port or add more ports here, you will likely also
+# have to change the VirtualHost statement in
+# /etc/apache2/sites-enabled/000-default.conf
+
+Listen 11111
```
Does this overlap with the ports used by the integration tests?

goatgoose replied:
Yes, it does! Thanks! I didn't think of that. I lowered this to 7777, which is below the minimum integration test port of 8000.

lrstewart commented on codebuild/bin/s2n_codebuild.sh:
```diff
@@ -65,6 +65,23 @@ if [[ -n "$S2N_NO_PQ" ]]; then
+source codebuild/bin/s2n_apache2.sh
+APACHE_CERT_DIR="$(pwd)/tests/pems"
+
+run_integration_v2_tests() {
+    apache2_start "${APACHE_CERT_DIR}"
```
Don't we usually run one integrationv2 test at a time in the CI? That would mean we'd spin apache up every time, even though only 1 out of like 18 tests needs it. Is there a way we can start apache only if the integration test will require it?

goatgoose replied:
That's true. Starting apache is fast, but it probably makes sense to check to see if it's needed first. I added a check that looks at the `TOX_TEST_NAME` variable, and enables it if the renegotiate test is being run.

goatgoose commented:
Docker image is rebuilding. Will update with codebuild run.

lrstewart commented on codebuild/bin/apache2/ports.conf:
```diff
@@ -2,6 +2,6 @@
 # /etc/apache2/sites-enabled/000-default.conf
 
-Listen 11111
+Listen 7777
```
I'd put a comment here noting that this should be below the integration test framework-- make sure someone else doesn't make the same mistake in the future :)

goatgoose replied:
Added!

lrstewart commented on codebuild/bin/apache2/sites-enabled/renegotiate.conf:
```diff
@@ -1,4 +1,4 @@
-<VirtualHost *:11111>
+<VirtualHost *:7777>
```
I'm not familiar with apache-- does this need to match the ports.conf? If so, why do we set the port in two different places?

goatgoose replied:
The virtual host port restricts the scope of the configuration to just one port, so it should match the port we tell apache to listen to in ports.conf. I made this a variable so we only set it once in ports.conf.

lrstewart commented on codebuild/bin/s2n_codebuild.sh:
```diff
@@ -65,18 +65,26 @@ if [[ -n "$S2N_NO_PQ" ]]; then
 
-source codebuild/bin/s2n_apache2.sh
-APACHE_CERT_DIR="$(pwd)/tests/pems"
+setup_integration_v2_tests() {
+    # Start the apache server if the list of tests isn't defined, meaning all tests
```
I'd be explicit in what this does: setup_apache_server

goatgoose replied:
Fixed!

lrstewart reviewed (commented):
Just nits now!

lrstewart commented on codebuild/bin/apache2/ports.conf:
```diff
@@ -0,0 +1,7 @@
+# Server ports should not conflict with the range of ports used for
+# integration tests (8000 to 30000)
+
+Define RENEGOTIATE_SERVER_PORT 7777
+Listen ${RENEGOTIATE_SERVER_PORT}
```
Nice :)

goatgoose commented:
Docker image is rebuilding. Will update with codebuild run.

lrstewart commented on codebuild/bin/s2n_codebuild.sh:
```diff
@@ -65,6 +65,31 @@ if [[ -n "$S2N_NO_PQ" ]]; then
+setup_apache_server() {
+    # Start the apache server if the list of tests isn't defined, meaning all tests
+    # are to be run, or if the renegotiate test is included in the list of tests.
+    if [[ -z $TOX_TEST_NAME ]] || [[ "${TOX_TEST_NAME}" == *"test_renegotiate"* ]]; then
+        source codebuild/bin/s2n_apache2.sh
```
Oh actually one more comment: why do you use $TOX_TEST_NAME here, but ${INTEGV2_TEST} on line 90? Is INTEGV2_TEST not used for the standard integrationv2 tests?

goatgoose replied:
For the standard integ tests, both variables should work in CI. Codebuild sets INTEGV2_TEST, which is set to TOX_TEST_NAME in the buildspec. However, for external users running our integ tests, our documentation says to set TOX_TEST_NAME, so this seems correct to check.

For criterion, it seems TOX_TEST_NAME isn't set until later, when calling make. Although, I think this line isn't being used currently, and it's removed in https://github.com/aws/s2n-tls/pull/3534 in favor of separate criterion buildspecs/scripts. To make this PR standalone, though, I added a test_names input to `setup_apache_server`.

goatgoose replied:
I actually think it makes more sense to just remove this here, and then there won't be conflicts when merging the criterion PR. 6ceadad96db7f3bb38fc0c65acb172976973362c

dougch commented on codebuild/bin/s2n_apache2.sh:
```diff
@@ -0,0 +1,43 @@
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+#
+
```
`set -eu` to make sure none of these are undefined?

goatgoose replied:
Didn't know about that, thanks! Added!

dougch commented on codebuild/bin/install_apache2.sh:
```diff
@@ -0,0 +1,82 @@
+# permissions and limitations under the License.
+#
+
+set -e
```
`set -eu` to catch undefined vars?

goatgoose replied:
Added!

dougch commented on codebuild/bin/s2n_codebuild.sh:
```diff
@@ -73,11 +91,9 @@ if [[ "$TESTS" == "ALL" || "$TESTS" == "unit" ]]; then cmake . -Bbuild -DCMAKE_P
 if [[ "$TESTS" == "ALL" || "$TESTS" == "asan" ]]; then make clean; S2N_ADDRESS_SANITIZER=1 make -j $JOBS ; fi
 if [[ "$TESTS" == "ALL" || "$TESTS" == "integration" ]]; then make clean; S2N_NO_SSLYZE=1 make integration ; fi
-if [[ "$TESTS" == "ALL" || "$TESTS" == "integrationv2" ]]; then $CB_BIN_DIR/install_s2n_head.sh "$(mktemp -d)"; make clean; make integrationv2 ; fi
+if [[ "$TESTS" == "ALL" || "$TESTS" == "integrationv2" ]]; then run_integration_v2_tests; fi
```
like this pattern

dougch reviewed (approved):
minor nits, cool stuff

dougch commented:
> New codebuild run: https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/S2nIntegrationV2Batch2/batch/S2nIntegrationV2Batch2%3Ac2385b09-b3fa-41d9-9c38-d78680ae9979/builds?region=us-west-2

Here's the [exact test_renegotiate successful child job](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/S2nIntegrationV2Batch2/build/S2nIntegrationV2Batch2%3A23c0ea33-1189-4c81-8577-c4c05f930dfb/log?region=us-west-2) for any other reviewer...

goatgoose commented:
I changed a few things to make running the integ tests outside of codebuild easier:
- Added a check for ubuntu before trying to install apache2:
https://github.com/aws/s2n-tls/blob/e1899df62ed0f07e27c1733b4440877af29d3bf5/codebuild/bin/install_default_dependencies.sh#L122-L132
- Before starting apache, it is checked to ensure apache is installed:
https://github.com/aws/s2n-tls/blob/e1899df62ed0f07e27c1733b4440877af29d3bf5/codebuild/bin/s2n_apache2.sh#L39-L43
- The renegotiation integ tests that communicate with apache will be in a separate file, `test_renegotiation_apache.py`. This allows the basic renegotiate tests to be run without installing apache:
https://github.com/aws/s2n-tls/blob/e1899df62ed0f07e27c1733b4440877af29d3bf5/codebuild/bin/s2n_codebuild.sh#L71

goatgoose commented:
Codebuild runs:

- [general batch](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/s2nGeneralBatch/batch/s2nGeneralBatch%3A8656efcb-85ce-4bcc-a576-8da91c236f59/details?region=us-west-2)
- [integ batch](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/S2nIntegrationV2Batch2/batch/S2nIntegrationV2Batch2%3Ae1a788d9-59a0-4e7e-8ab7-e16f23787a80?region=us-west-2)
